server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java

```diff
@@ -68,21 +68,29 @@
         }
         this.maxParentPathDots = maxParentPathDots;
 
+        Map<String, MappedFieldType> resolvedRuntimeMappers = new HashMap<>();
+        Map<String, DynamicFieldType> resolvedDynamicMappers = new HashMap<>();
         for (FieldAliasMapper fieldAliasMapper : fieldAliasMappers) {
             String aliasName = fieldAliasMapper.name();
             String path = fieldAliasMapper.path();
-            MappedFieldType fieldType = fullNameToFieldType.get(path);
-            fullNameToFieldType.put(aliasName, fieldType);
-            if (fieldType instanceof DynamicFieldType) {
-                dynamicFieldTypes.put(aliasName, (DynamicFieldType) fieldType);
+            RuntimeField aliasField = new AliasRuntimeField(aliasName, path);
+            MappedFieldType resolved = aliasField.asMappedFieldType(fullNameToFieldType::get);
+            resolvedRuntimeMappers.put(aliasName, resolved);
+            if (resolved instanceof DynamicFieldType) {
+                resolvedDynamicMappers.put(aliasName, (DynamicFieldType) resolved);
             }
         }
 
         for (RuntimeField runtimeField : runtimeFields) {
-            MappedFieldType runtimeFieldType = runtimeField.asMappedFieldType();
-            //this will override concrete fields with runtime fields that have the same name
-            fullNameToFieldType.put(runtimeFieldType.name(), runtimeFieldType);
+            MappedFieldType resolved = runtimeField.asMappedFieldType(fullNameToFieldType::get);
```

Comment:
thinking out loud: if we use asMappedFieldType only for aliases, given that all existing runtime fields return `this`, should we consider adding the notion of alias to RuntimeField through a method that returns a String and return by default the same as the field name, and when it differs look it up directly in FieldTypeLookup? Maybe it is cryptic, not too sure, but it would make it more specific to aliases (which is both good and bad I guess), would remove the need to delegate to the runtime field to look up the aliased field type. I do wonder if it should be the responsibility of the field type to look up the aliases field among all fields. We may or may not need something more generic at some point in the future but we can always refactor things later when needed?

Reply:
one more thing: fullNameToFieldType does not include the previously resolved alias field mapper, so we are sure that an alias runtime field never points to an alias field defined under properties?